### DISCUSSION TITLE
ApplicationContext.getBeansOfType() returning different number of results in test

### DISCUSSION THREAD

**Discussion by cancech** [micronaut-projects/micronaut-core#8308] [Q&A] [answered]
In our application we have an `Initializer` interface which classes can implement if they require to be initialized on startup. The idea being that as part of the application startup the following is called

```
@Inject
ApplicationContext ctx;

ctx.getBeansOfType(Initializer.class).forEach(Initializer::initialize);
```

Thus in a test I do the following

```
@Inject
ApplicationContext ctx;

@BeforeEach
public void setup() {
    ctx.getBeansOfType(Initializer.class).forEach(Initializer::initialize);
}
```

However, what I'm seeing is that the number of `Initializer` beans retrieved at the start of the test differs from execution to execution. Basically, when I run the test I'm getting anywhere between 20 - 35 `Initializer` instances retrieved/initialized. Why would this number fluctuate? I'm not changing any code or anything else from run to run, so I would expect the number of retrieved instances to always be the same. Not to mention that some instances only ever seem to be retrieved when I explicitly `@Inject` it into the test. I started down this rabbit hole when testing to make sure that that multiple copies of a concrete `Initialzier` are created/initialized when creating it using a `@EachProperty`, `@EachBean` combo. In the test I'm trying to create two copies of the concrete `Initializer`, and the first is always retrieved/initialized, while the second only when I explicitly `@Inject` it in the test.

Note: `Initializer` instances are being created in basically every imaginable manner (as Lambdas via `@Factory` factory, concrete bean that implements the interface, as as well duplication as indicated above).

**Comment by graemerocher** [comment, marked as answer]
Hard to say without a reproducer but sounds like a bug for sure

**Reply by cancech** [reply]
I've created and attached a little contrived project where this behavior can be seen (I'm using OpenJDK11.0.2, Gradle 7.2.0). It' just creating a number of beans which implement a `MyPrinter` interface and then triggering them on initialization. There are 30 such instances created, and all thirty are loaded/processed when running the application proper. However when running the test, only 20 of the total 30 are loaded.

[micronaut-bug.zip](https://github.com/micronaut-projects/micronaut-core/files/9971195/micronaut-bug.zip)

At this point the only thing that I can think of, perhaps a threading issue? If the `ApplicationContext` is created in a separate thread from the initialization of the test, then perhaps the tests are triggered before the `ApplicationContext` finishes loading all of the beans?

**Reply by graemerocher** [reply]
could you create an issue report https://github.com/micronaut-projects/micronaut-core/issues

**Reply by cancech** [reply]
Issue has been created: #8315

Marking this as answered as anything further will be handled in the issue report.

Thanks!
